File: 01. DataStructures/02. Stack/LeetCode Questions/09. Valid Parentheses #20/Solution.java
class Solution {
    public boolean isValid(String s) {
        if(s.length()%2!=0) return false;
        
        Stack<Character> st = new Stack();
       for(char c :  s.toCharArray()){ 
               if(c=='(' || c=='[' ||  c=='{')
                   st.push(c);
               else if(c==']' && !st.isEmpty() &&  st.peek()=='[') st.pop();
               else if(c==')' && !st.isEmpty() &&  st.peek()=='(') st.pop();
               else if(c=='}' && !st.isEmpty() &&  st.peek()=='{') st.pop();
               else st.push(c);
        }
       return st.isEmpty();
    }
}

File: 01. DataStructures/02. Stack/LeetCode Questions/09. Valid Parentheses #20/solution.js
var isValid = function(s) {
    
    paranthesesMap = new Map();
    paranthesesMap.set(')', '(');
    paranthesesMap.set('}', '{');
    paranthesesMap.set(']', '[');

    validationStack = [];
    validationStack[0] = s[0];
    
    for(let i = 1; i < s.length; i++) {
        if(validationStack[validationStack.length - 1] && validationStack[validationStack.length - 1] === paranthesesMap.get(s[i])) {
            validationStack.pop();
        } else {
            validationStack.push(s[i]);
        }
    }
    return validationStack.length > 0 ?  false: true;
};
// {
console.log(isValid("((())){[]}[()](())([])({}){()}"));
